pkg/destroy/azure/azure.go

```diff
@@ -0,0 +1,247 @@
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/dns/mgmt/2018-03-01-preview/dns"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	session "github.com/openshift/installer/pkg/asset/installconfig/azure"
+	"github.com/openshift/installer/pkg/destroy"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// ClusterUninstaller holds the various options for the cluster we want to delete.
+type ClusterUninstaller struct {
+	SubscriptionID string
+	Authorizer     autorest.Authorizer
+
+	InfraID string
+
+	Logger logrus.FieldLogger
+
+	resourceGroupsClient resources.GroupsGroupClient
+	zonesClient          dns.ZonesClient
+	recordsClient        dns.RecordSetsClient
+}
+
+func (o *ClusterUninstaller) configureClients() {
+	resourceGroupsClient := resources.NewGroupsGroupClient(o.SubscriptionID)
+	resourceGroupsClient.Authorizer = o.Authorizer
+	o.resourceGroupsClient = resourceGroupsClient
+
+	zonesClient := dns.NewZonesClient(o.SubscriptionID)
+	zonesClient.Authorizer = o.Authorizer
+	o.zonesClient = zonesClient
+
+	recordsClient := dns.NewRecordSetsClient(o.SubscriptionID)
+	recordsClient.Authorizer = o.Authorizer
+	o.recordsClient = recordsClient
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (o *ClusterUninstaller) Run() error {
+	o.configureClients()
+	group := o.InfraID + "-rg"
+	o.Logger.Debug("deleting public records")
+	if err := deletePublicRecords(o.zonesClient, o.recordsClient, o.Logger, group); err != nil {
+		o.Logger.Debug(err)
+		return errors.Wrap(err, "failed to delete public dns records")
+	}
+	o.Logger.Debug("deleting resource group")
+	if err := deleteResourceGroup(o.resourceGroupsClient, o.Logger, group); err != nil {
+		o.Logger.Debug(err)
+		return errors.Wrap(err, "failed to delete resource group")
+	}
+
+	return nil
+}
+
+func deletePublicRecords(dnsClient dns.ZonesClient, recordsClient dns.RecordSetsClient, logger logrus.FieldLogger, rgName string) error {
+	logger = logger.WithField("resource group", rgName)
+	ctx, cancel := context.WithTimeout(context.TODO(), 10*time.Minute)
+	defer cancel()
+
+	// collect records from private zones in rgName
+	allPrivateZonesInGroup := []string{}
+	for zonesPage, err := dnsClient.ListByResourceGroup(ctx, rgName, to.Int32Ptr(100)); zonesPage.NotDone(); err = zonesPage.NextWithContext(ctx) {
+		if err != nil {
+			return err
+		}
+		for _, zone := range zonesPage.Values() {
+			if zone.ZoneType == dns.Private {
+				allPrivateZonesInGroup = append(allPrivateZonesInGroup, to.String(zone.Name))
+				continue
+			}
+		}
+	}
+
+	var errs []error
+	for _, zone := range allPrivateZonesInGroup {
+		if err := deletePublicRecordsForZone(ctx, dnsClient, recordsClient, logger, rgName, zone); err != nil {
```

Comment:
Just call this from the earlier loop without bothering with `allPrivateZonesInGroup`?

Reply:
done with https://github.com/openshift/installer/pull/1771/commits/3d25470009039e20339d645c8a4454559406e914 -> https://github.com/openshift/installer/pull/1771/commits/27da87f528d10f6dd1d7678e7ec859365c1df300